### PR TITLE
Add support for Disable field in generate blocks

### LINKED ISSUE #2496: Support for disabling generate blocks
I have a use case where it would be helpful to disable generate blocks on a per environment basis. I work with a repository that contains dev environments that have their terraform source pointed to a local path and prod environments that have their source pointed to a git tag. I often run into a problem when adding a new generate block where the dev environments work as expected but the prod environment fail because a dependency has not yet been added to a prod release.

Ideally I would like to be able to introduce a new generate block to our dev environments first and have it disabled in prod using `local` vars per environment.

### PR DESCRIPTION
**Pull request by szesch**
## Description

Fixes #2496.

Introduces a `disable` field in `generate` blocks so they can be disabled on a per use case basis.

## TODOs

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added 'disable' field to `generate` blocks.

### CODE REVIEW

**Comment by denis256** [comment]
Noticed that some of integration tests failed:

```
TestRenderJSONConfig

=== CONT  TestRenderJSONConfig
    integration_debug_test.go:246: 
        	Error Trace:	integration_debug_test.go:246
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"provider \"aws\" {\n  region = \"us-east-1\"\n}\n", "disable_signature":false, "if_exists":"overwrite_terragrunt", "path":"provider.tf"}}
        	            	actual  : map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"provider \"aws\" {\n  region = \"us-east-1\"\n}\n", "disable":false, "disable_signature":false, "if_exists":"overwrite_terragrunt", "path":"provider.tf"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,5 +1,6 @@
        	            	 (map[string]interface {}) (len=1) {
        	            	- (string) (len=8) "provider": (map[string]interface {}) (len=5) {
        	            	+ (string) (len=8) "provider": (map[string]interface {}) (len=6) {
        	            	   (string) (len=14) "comment_prefix": (string) (len=2) "# ",
        	            	   (string) (len=8) "contents": (string) (len=42) "provider \"aws\" {\n  region = \"us-east-1\"\n}\n",
        	            	+  (string) (len=7) "disable": (bool) false,
        	            	   (string) (len=17) "disable_signature": (bool) false,
        	Test:       	TestRenderJSONConfig
--- FAIL: TestRenderJSONConfig (293.25s)


TestRenderJSONConfigWithIncludesDependenciesAndLocals

=== CONT  TestRenderJSONConfigWithIncludesDependenciesAndLocals
    integration_debug_test.go:345: 
        	Error Trace:	integration_debug_test.go:345
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"# This is just a test", "disable_signature":false, "if_exists":"overwrite", "path":"provider.tf"}}
        	            	actual  : map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"# This is just a test", "disable":false, "disable_signature":false, "if_exists":"overwrite", "path":"provider.tf"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,5 +1,6 @@
        	            	 (map[string]interface {}) (len=1) {
        	            	- (string) (len=8) "provider": (map[string]interface {}) (len=5) {
        	            	+ (string) (len=8) "provider": (map[string]interface {}) (len=6) {
        	            	   (string) (len=14) "comment_prefix": (string) (len=2) "# ",
        	            	   (string) (len=8) "contents": (string) (len=21) "# This is just a test",
        	            	+  (string) (len=7) "disable": (bool) false,
        	            	   (string) (len=17) "disable_signature": (bool) false,
        	Test:       	TestRenderJSONConfigWithIncludesDependenciesAndLocals
--- FAIL: TestRenderJSONConfigWithIncludesDependenciesAndLocals (254.76s)


TestReadTerragruntConfigFull
=== CONT  TestReadTerragruntConfigFull
    integration_test.go:3286: 
        	Error Trace:	integration_test.go:3286
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"provider \"aws\" {\n  region = \"us-east-1\"\n}\n", "disable":false, "disable_signature":false, "if_exists":"overwrite_terragrunt", "path":"provider.tf"}}
        	            	actual  : map[string]interface {}{"provider":map[string]interface {}{"comment_prefix":"# ", "contents":"provider \"aws\" {\n  region = \"us-east-1\"\n}\n", "disable_signature":false, "if_exists":"overwrite_terragrunt", "path":"provider.tf"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,6 +1,5 @@
        	            	 (map[string]interface {}) (len=1) {
        	            	- (string) (len=8) "provider": (map[string]interface {}) (len=6) {
        	            	+ (string) (len=8) "provider": (map[string]interface {}) (len=5) {
        	            	   (string) (len=14) "comment_prefix": (string) (len=2) "# ",
        	            	   (string) (len=8) "contents": (string) (len=42) "provider \"aws\" {\n  region = \"us-east-1\"\n}\n",
        	            	-  (string) (len=7) "disable": (bool) false,
        	            	   (string) (len=17) "disable_signature": (bool) false,
        	Test:       	TestReadTerragruntConfigFull
--- FAIL: TestReadTerragruntConfigFull (170.10s)

TestRenderJsonMetadataIncludes

=== CONT  TestRenderJsonMetadataIncludes
    integration_test.go:4958: 
        	Error Trace:	integration_test.go:4958
        	Error:      	Not equal: 
        	            	expected: "{\"provider\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test4058810509/fixture-render-json-metadata/includes/app/generate.hcl\"},\"value\":{\"comment_prefix\":\"# \",\"contents\":\"# test\\n\",\"disable_signature\":false,\"if_exists\":\"overwrite\",\"path\":\"provider.tf\"}}}"
        	            	actual  : "{\"provider\":{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test4058810509/fixture-render-json-metadata/includes/app/generate.hcl\"},\"value\":{\"comment_prefix\":\"# \",\"contents\":\"# test\\n\",\"disable\":false,\"disable_signature\":false,\"if_exists\":\"overwrite\",\"path\":\"provider.tf\"}}}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-{"provider":{"metadata":{"found_in_file":"/tmp/terragrunt-test4058810509/fixture-render-json-metadata/includes/app/generate.hcl"},"value":{"comment_prefix":"# ","contents":"# test\n","disable_signature":false,"if_exists":"overwrite","path":"provider.tf"}}}
        	            	+{"provider":{"metadata":{"found_in_file":"/tmp/terragrunt-test4058810509/fixture-render-json-metadata/includes/app/generate.hcl"},"value":{"comment_prefix":"# ","contents":"# test\n","disable":false,"disable_signature":false,"if_exists":"overwrite","path":"provider.tf"}}}
        	Test:       	TestRenderJsonMetadataIncludes
--- FAIL: TestRenderJsonMetadataIncludes (22.36s)

```

**Comment by szesch** [comment]
@denis256 I've updated the tests.

**Comment by szesch** [comment]
@denis256 is there anything else needed?

**Comment by ivankatliarchuk** [comment]
Have you considered the `enabled` input parameter instead of `disabled`? In some cases, the generator block may need to be enabled or disabled based on certain conditions.

Which one is easier to read and understand?

```
enabled = true|false
disabled = false|true`
```

**Comment by szesch** [comment]
My original plan was to use `enabled` but after reviewing other configuration I noticed there is an existing pattern that uses `disable*`. For instance, `remote_state` has `disable_init`, `disable_dependency_optimization`, `disable_bucket_update`, etc. `generate` has `disable_signature`. So I ultimately went with `disabled` to keep it consistent.

**Comment by szesch** [comment]
@denis256 @zackproser can this be merged? Thanks!

File: test/fixture-codegen/generate-block/disable/terragrunt.hcl [hunk 1]
```diff
@@ -0,0 +1,10 @@
+generate "backend" {
+  path = "data.txt"
+  contents = "test data1"
+  if_exists = "error"
+  disable = true
```
[the remark this answers is not in the document]
Comment (szesch) [reply test/fixture-codegen/generate-block/disable/terragrunt.hcl:5]: I've pushed an additional test

**Comment by szesch** [comment]
@denis256 @zackproser  Please take a look and let me know if anything else is required. Thanks!

**Comment by denis256** [comment]
Hello,
CICD job is failing now, I think because branch is behind `gruntwork-io:master`

https://github.com/szesch/terragrunt/tree/feature/disable-generate-blocks
```
This branch is 3 commits ahead, 21 commits behind gruntwork-io:master.
```

**Review by denis256 — APPROVED** [review]

**Pull request merged**